### PR TITLE
New feature extractor

### PR DESCRIPTION
**Pull request by ShunsukeOnoo**
I fixed an issue that `FeatureExtractor` can return features mismatched keys and values. 
This issue is because `FeatureExtractor` assumes that the features are stored in `FeatureExtractorHandle` in the same order as the layers, but in reality, they are ordered based on the sequence in which each layer is called.
As a fix, the new hook class `_Hook` stores features in a `dict` instead of `list`, mitigating the problem.

### CODE REVIEW

File: bdpy/dl/torch/torch.py [hunk 1]
```diff
@@ -17,128 +17,108 @@
...
+            self._hook_handler[name] = hook_handle  # Store the actual hook handle
+    
+    def __call__(self, x: Union[torch.Tensor, np.ndarray]) -> Dict[str, Union[torch.Tensor, np.ndarray]]:
         return self.run(x)
-
-    def run(self, x: _tensor_t) -> Dict[str, np.ndarray] | Dict[str, torch.Tensor]:
-        '''Extract feature activations from the specified layers.
-
+    
+    def run(self, x: Union[torch.Tensor, np.ndarray]) -> Dict[str, Union[torch.Tensor, np.ndarray]]:
```
Comment (ganow) [inline bdpy/dl/torch/torch.py:86]: Can you change the type annotation of `.run(x)` from

```
Union[Tensor, ndarray] -> Dict[str, Union[Tensor, ndarray]]
```

to

```
Union[Tensor -> Dict[str, Tensor], ndarray -> Dict[str, ndarray]]
```

?

This can be achieved by `TypeVar`. Here is the minimal example:

```python
import torch
import numpy as np
from typing import TypeVar

_T = TypeVar("_T", torch.Tensor, np.ndarray)

def just_return_either(x: _T) -> _T:
    return x

x = torch.tensor(1.0)
y = just_return_either(x)  # inferred as torch.Tensor instead of Union[torch.Tensor, np.ndarray]

x_np = np.array(1.0)
y_np = just_return_either(x_np)  # inferred as np.ndarray instead of Union[torch.Tensor, np.ndarray]
```

**Review by ganow — CHANGES_REQUESTED** [review]

File: bdpy/dl/torch/torch.py [hunk 2]
```diff
@@ -17,128 +17,108 @@
...
-            module: nn.Module, module_in: Any,
-            module_out: torch.Tensor
-    ) -> None:
-        self.outputs.append(module_out.detach().clone())
-
-    def clear(self):
-        self.outputs = []
+        """Removes all registered hooks when the instance is deleted."""
+        for name, hook in self._hook_handler.items():
+            hook.remove()
```
[the remark this answers is not in the document]
Comment (ganow) [reply bdpy/dl/torch/torch.py:120]: oh I'm sorry now I'm realized that the `hook` object here is actually the `hook_handler` object. Could you refactor the variable names as the following?

- `self._hook_handler` -> `self._hook_handlers`
- `hook` in this loop -> `hook_handler`

File: bdpy/dl/torch/torch.py [hunk 3]
```diff
@@ -17,128 +17,108 @@
...
+        dict[str, Union[torch.Tensor, np.ndarray]]
+            A dictionary containing features extracted from the specified layers, 
+            optionally transformed.
+        """
         if not isinstance(x, torch.Tensor):
-            xt = torch.tensor(x[np.newaxis], device=self.__device)
-        else:
-            xt = x
+            # TODO: From legacy code. Is this necessary?
+            x = torch.tensor(x[np.newaxis], device=self.device)
```
Comment (ganow) [inline bdpy/dl/torch/torch.py:103]:
```suggestion
            warnings.warn(
                "In future versions, it will no longer be possible to input ndarrays " \
                "directly into FeatureExtractor. Please convert the input to torch.Tensor.",
                DeprecationWarning,
                stacklevel=2
            )
            x = torch.tensor(x[np.newaxis], device=self.device)
```

File: bdpy/dl/torch/torch.py [hunk 4]
```diff
@@ -17,128 +17,108 @@
...
 
+
+class FeatureExtractor:
+    def __init__(
+            self, 
+            encoder: nn.Module, 
+            layers: List[str], 
+            layer_mapping: Optional[Dict[str, str]] = None, 
+            device: str = 'cpu',
+            detach: bool = False,
```
[the remark this answers is not in the document]
Comment (ShunsukeOnoo) [reply bdpy/dl/torch/torch.py:38]: I unintentionally changed that. I will change it back to `True`.

File: bdpy/dl/torch/torch.py (hunk 1 above)
Comment (ShunsukeOnoo) [reply bdpy/dl/torch/torch.py:86]: Do you mean the input and return type should be matched, (if `np.ndarray` is passed, then the return type should be `Dict[str, np.ndarray]`)? If so, it will be a big change in the behavior because in the previous implementation, return type is determined by `detach` parameter rather than the input type.

Comment (ganow) [reply bdpy/dl/torch/torch.py:86]: I'm sorry I was confused. You are right, the return type of `FeatureExtractor.run(x)` should depends on the initializing variable `detach` instead of the input argument type `type(x)`. So, could you implement in a way like this?

```python
from typing import Union, overload, Literal, Dict, TypeVar, Generic, TypeAlias
import torch
import numpy as np

_tensor_t: TypeAlias = Union[np.ndarray, torch.Tensor]
_T = TypeVar('_T', bound=Union[np.ndarray, torch.Tensor])

class FeatureExtractor(Generic[_T]):
    @overload
    def __init__(self: "FeatureExtractor[torch.Tensor]") -> None: ...

    @overload
    def __init__(self: "FeatureExtractor[np.ndarray]", detach: Literal[True]) -> None: ...

    @overload
    def __init__(self: "FeatureExtractor[torch.Tensor]", detach: Literal[False]) -> None: ...

    def __init__(self, detach = False):
        self.detach = detach

    def run(self, x: _tensor_t) -> Dict[str, _T]:
        features: Dict[str, torch.Tensor] = ...
        if self.detach:
            return {
                k: v.cpu().detach().numpy()
                for k, v in features.items()
            }  # type: ignore
        else:
            return features  # type: ignore
```

By using the above annotation, the results would be like:

```python
f1 = FeatureExtractor(detach=True)
y1 = f.run(np.array([1, 2, 3]))  # inferred as dict[str, np.ndarray]
f2 = FeatureExtractor(detach=False)
y2 = f2.run(np.array([1, 2, 3]))  # inferred as dict[str, torch.Tensor]
f3 = FeatureExtractor()
y3 = f3.run(torch.tensor([1, 2, 3]))  # inferred as dict[str, torch.Tensor]
```
